frame/session/src/historical.rs

```diff
@@ -253,26 +254,16 @@ impl<T: Trait> ProvingTrie<T> {
 
 }
 
-/// Proof of ownership of a specific key.
-#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
-pub struct Proof {
-	session: SessionIndex,
-	trie_nodes: Vec<Vec<u8>>,
-}
-
-impl Proof {
-	/// Returns a session this proof was generated for.
-	pub fn session(&self) -> SessionIndex {
-		self.session
-	}
-}
-
 impl<T: Trait, D: AsRef<[u8]>> frame_support::traits::KeyOwnerProofSystem<(KeyTypeId, D)>
 	for Module<T>
 {
-	type Proof = Proof;
+	type Proof = MembershipProof;
 	type IdentificationTuple = IdentificationTuple<T>;
 
+	// The session index of a given membership proof and the validator count for
+	// that session.
+	type ExtraData = (SessionIndex, ValidatorCount);
```

Comment:
This approach seems difficult to reconcile with the trait system. The property that users of `KeyOwnerProofSystem` need would be something like `Proof::ExtraData: Has<SessionIndex> + Has<ValidatorCount>` which can only be expressed via `frunk`'s `HList` ATM. Either that or you require `ExtraData = (SessionIndex, ValidatorCount)` which feels fragile.

The way it was done for parachain-equivocations in Polkadot was to add a trait and then you can just ensure `Proof: GetSessionIndex + GetValidatorCount`

Reply:
Implemented this now and reverted the changes to the historical module.